File: pycodes/backtracking.py
import copy
class Solution(object):
    # 77. 组合
    # 我的，写的有点乱八七糟，答案也是这样，剪枝过，我真厉害
    def combine(self, n, k):
        res = []
        def func(low, n, k ,choice,res):
            if k == 0:
                res.append(choice)
                return
            for i in range(low,n-k+2):
                choice.append(i)
                func(i+1,n,k-1,choice[:],res)
                choice.pop()
        func(1,n,k,[],res)
        return res

    # 216. 组合总和 III
    def combinationSum3(self, k, n):
        res = []
        def func(k,n,low,path,res):
            total = sum(path)
            if k == 0 and total == n:
                res.append(path)
            elif k > 0 and total < n:
                for i in range(low, 11-k):
                    path.append(i)
                    func(k-1,n,i+1,path[:],res)
                    path.pop()
            else:return
        func(k, n, 1, [], res)
        return res
    
    # 17. 电话号码的字母组合
        def letterCombinations(self, digits):
            if len(digits) == 0: return []
            res = []
            dic = {"2":["a", "b", "c"], "3":["d","e","f"], "4":["g", "h", "i"],
                   "5":["j","k","l"], "6":["m","n","o"], "7":["p","q","r","s"], 
                   "8":["t","u","v"], "9":["w","x","y","z"]}
            def func(digits, combine, res, d):
                if len(digits)==0: 
                    res.append(combine)
                    return
                else:
                    for char in d[digits[0]]:
                        func(digits[1:], combine+char, res, d)
            func(digits, "", res, dic)
            return res

    # 39. 组合总和
    def combinationSum(self, candidates, target):
        res = []
        def func(cand, pos, tar, path, cursum, res):
            if cursum == tar: 
                res.append(path)
                return
            elif cursum > tar: return
            else:
                for i in range(pos,len(cand)):
                    if cand[i] + cursum <= tar:
                        func(cand, i, tar, path+[cand[i]], cursum + cand[i], res)
                    else: break
        func(candidates.sort(), 0, target, [], 0, res)
        return res
    
    # 40.组合总和II    
    def combinationSum2(self, candidates, target):
        candidates.sort()
        res = []
        def func(cand, pos, tar, path, cursum, res):
            if cursum == tar:
                res.append(path)
                return
            elif cursum > tar: return
            else:
                i = pos
                while i < len(cand):
                    if cand[i] + cursum <= tar:
                        func(cand, i+1, tar, path+[cand[i]], cursum+cand[i], res)
                        newi = i + 1
                        while newi < len(cand):
                            if cand[i] == cand[newi]: newi+=1
                            else:break
                        i = newi
                    else:break
        func(candidates, 0, target, [], 0, res)
        return res

    # 131. 分割回文串
    # 偶写的基础班，妹有剪枝
    def partition(self, s):
        res = []
        def ispalindrome(s):
            left, right = 0, len(s)-1
            while left < right:
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1
            return True
        def backtracking(s,path,res):
            if s == "":res.append(path)
            else:
                for i in range(1, len(s)+1):
                    if ispalindrome(s[0:i]):
                        backtracking(s[i:], path[:]+[s[0:i]], res)
        backtracking(s,[],res)
        return res
    
    # 93. 复原 IP 地址
    def restoreIpAddresses(self, s):
        res = []
        def backtracking(address, path, n, res):
            if n == 0:
                if address == "":
                    res.append(".".join(path))
                return 
            if len(address) > 3*n or len(address) < n:
                return
            else:
                if address[0] == "0":
                    backtracking(address[1:], path+[address[0]], n-1, res)
                else:
                    for i in range(1,min(len(address)+1,4)):
                        if int(address[:i]) <= 255:
                            backtracking(address[i:], path+[address[:i]], n-1, res)
        backtracking(s,[],4,res)
        return res
         
    # 78. 子集
    def subsets(self, nums):
        res = []
        def backtracking(nums, res, path):
            if len(nums) == 0:
                res.append(path)
                return
            else:
                backtracking(nums[1:], res, path[:])
                backtracking(nums[1:], res, path[:]+nums[:1])
        backtracking(nums, res, [])
        return res

    def subsets1(self, nums):
        res = []
        def backtracking(nums, res, path):
            res.append(path)
            for i in range(len(nums)):
                backtracking(nums[i+1:], res, path+[nums[i]])
        backtracking(nums, res, [])
        return res
    
    # 79. 子集二    
    def subsetsWithDup(self, nums):
        nums.sort()
        res = []
        def backtracking(nums, res, path, startindex):
            res.append(path)
            cur = None
            for i in range(startindex, len(nums)):
                if nums[i] != cur:
                    backtracking(nums, res, path+[nums[i]], i+1)
                    cur = nums[i]
        backtracking(nums, res, [], 0)
        return res 
    
    # 491. 递增子序列
    # 注意这一道题和上一道题的去重方法不一样，上一道题的nums是进行排序了的，所以for循环的时候如果num[i]使用了一次之后，只需要用cur记录下来，
    # 后面比它大的数是一定没有使用的，从而达到去重目的。
    # 但是这题的nums是不能进行排序的，所以后面遇到的元素可能为for循环之前使用过的任一元素，因此要用集合把使用过的元素保存下来。
    def findSubsequences(self, nums):
        res = []
        def backtracking(nums, res, path, pre):
            if len(path) >= 2:
                res.append(path)
            used = set()
            for i in range(len(nums)):
                if nums[i] not in used and nums[i] >= pre:
                    backtracking(nums[i+1:], res, path+[nums[i]], nums[i])
                    used.add(nums[i])

        backtracking(nums, res, [], float('-inf'))
        return res
    
    # 46. 全排列
    def permute(self, nums):
        res = []
        path = [None]*len(nums)
        def backtracking(nums, path, res):
            if len(nums) == 0:
                res.append(path)
                return
            for pos in range(len(path)):
                if path[pos] == None:
                    path[pos] = nums[0]
                    backtracking(nums[1:], path[:], res)
                    path[pos] = None
        backtracking(nums, path[:], res)
        return res
    # 47. 全排列2
    def permuteUnique(self, nums):
        nums.sort()
        res = []
        def backtracking(nums, used, path, res):
            if len(path) == len(nums):
                res.append(path)
                return
            cur = None
            for i in range(len(nums)):
                if not used[i] and nums[i] != cur:
                    used[i] = True
                    backtracking(nums, used, path[:]+[nums[i]], res)
                    used[i] = False
                    cur = nums[i]
        backtracking(nums, [False]*len(nums), [], res)
        return res
    
    # 332. 重新安排行程
    # 注意这样写找了所有可以使用的行程，但是题目只需要一个就可以了
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        res = []
        def backtracking(tickets, used, cur_pos, res, path):
            if len(path) == len(tickets) + 1:
                res.append(path)
                return True
            for i in range(len(tickets)):
                if not used[i] and tickets[i][0] == cur_pos:
                    used[i] = True
                    find = backtracking(tickets, used[:], tickets[i][1], res, path+[tickets[i][1]])
                    used[i] = False
                    if find: return True
            return res
        tickets.sort()
        backtracking(tickets, [False]*len(tickets), "JFK", res, ["JFK"])
        return res[0]

    # 51. N 皇后
    # chessboard = [["."]*n]*n
    def solveNQueens(self, n):
        if n == 1: return [["Q"]]
        if n <= 3: return []
        res = []
        chessboard = [["." for i in range(n)] for j in range(n)]
        def check(pos, chessboard, n):
            x,y = pos
            for i in range(n):
                if chessboard[x][i] == "Q" or chessboard[i][y] == "Q":
                    return False
            a, b = (x-y, 0) if x > y else (0, y-x)
            while a < n and b < n:
                if chessboard[a][b] == "Q":
                    return False
                a += 1
                b += 1
            if x + y < n:
                a, b = (x+y, 0)
            else:
                a, b = (n-1,x+y+1-n)
            while a >= 0 and b < n:
                if chessboard[a][b] == "Q":
                     return False
                a -= 1
                b += 1
            return True
        def backtracking(n, res, chessboard, line):
            if line >= n:
                res.append(["".join(copy.deepcopy(chessboard)[i]) for i in range(n)])
                return
            else:
                for i in range(n):
                    if check((line, i), chessboard, n):
                        chessboard[line][i] = "Q"
                        backtracking(n, res, chessboard, line+1)
                        chessboard[line][i] = "."
        
        backtracking(n, res, chessboard, 0)
        return res


    # 37. 解数独
    def solveSudoku(self, board):
        def check(x, y, board, num):
            for i in range(9):
                if board[x][i] == num or board[i][y] == num:
                    return False
            x = x - x%3
            y = y - y%3
            for i in range(3):
                for j in range(3):
                    if board[x+i][y+j] == num:
                        return False
            return True
        def backtracking(x,y,board):
            if x == 9: return True
            if y == 8:
                nextx, nexty = x+1, 0
            else:
                nextx, nexty = x, y+1
            if board[x][y] == ".":
                for num in range(1,10):
                    if check(x,y,board,str(num)):
                        board[x][y] = str(num)
                        if backtracking(nextx, nexty, board): return True
                        board[x][y] = "."
            else:
                if backtracking(nextx, nexty, board): return True
        backtracking(0,0,board)                    

File: cpp/code/backtracking.cpp
# include<iostream>
# include<string>
# include<algorithm>
# include<limits>
# include<queue>
# include<vector>
# include<stack>
# include<list>
# include <map>
#include <unordered_map>

using namespace std;

class Solution_77{
public:
    
    // 77 组合
    void backtracking_combine(vector<vector<int>>& res, int n, int k, int low, vector<int> path){
        int lenPath = path.size();
        if(lenPath == k) {
            res.push_back(path);
            return;
        }
        for(int i = low; i <= n - (k - 1 - lenPath); i++){
            path.push_back(i);
            backtracking_combine(res, n, k, i + 1, path);
            path.pop_back();
        }
    }

    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        backtracking_combine(res, n, k, 1, vector<int>());
        return res;
    }

};


// 216. 组合总和 III
class Solution_216{
private:
    vector<vector<int>> res;
    vector<int> path;
    void backtracking(int n, int k, int begin){
        if(k == 0){
            if(n == 0) res.push_back(path);
            return;
        }
        if(n <= 0) return;
        for(int i = begin; i <= 10-k; i++){
            path.push_back(i);
            backtracking(n - i, k - 1, i + 1);
            path.pop_back();
        }
    }

public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        backtracking(n, k, 0);
        return res;
    }
};

class Solution{
private:
    map<char, string> digits_17{{'2',"abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};

public:
    // 17. 电话号码的字母组合
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        backtracking_17(res, digits, 0, "");
        return res;
    }

    void backtracking_17(vector<string>& res, string digits, int index, string path){
        if(digits.length() == 0) return;
        if(index == digits.length()) res.push_back(path);
        for(char ch: digits_17[digits[index]]){
            backtracking_17(res, digits, index+1, path+ch);
        }
    }

    // 39. 组合总和
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        backtracking_39(res, 0, candidates, target, vector<int>());
        return res;
    }

    void backtracking_39(vector<vector<int>> res, int index, vector<int>& candidates, int target, vector<int> path){
        if(target == 0){
            res.push_back(path);
            return;
        }
        if(candidates[index] > target) return;
        for(int i = index; i < candidates.size(); i++){
            path.push_back(candidates[i]);
            backtracking_39(res, i, candidates, target-candidates[i], path);
            path.pop_back();
        }
    }
};


// 40. 组合总和 II
class Solution_40_1 {
private:
    vector<int> path;
    vector<vector<int>> res;
public:
    void backtracking(vector<int>& candidates, int index, int target){
        if(target == 0){
            res.push_back(path);
            return;
        }
        for(int i = index; i < candidates.size() && candidates[i] <= target; i++){
            if(i!=0 and candidates[i]==candidates[i-1]) continue;
            path.push_back(candidates[i]);
            backtracking(candidates, i + 1, target - candidates[i]);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<bool> used(candidates.size(), false);
        sort(candidates.begin(), candidates.end());
        backtracking(candidates, 0, target);
        return res;
    }
};

// 131分割回文串
class Solution_131 {
private:
    vector<vector<string>> res;
    vector<string> path;
public:
    bool isPalindrome(string s, int begin, int end){
        while(begin<end){
            if(s[begin++] != s[end++]) return false;
        }
        return true;
    }
    vector<vector<string>> partition(string s) {
        backtracking(s, 0);
        return res;
    }
    void backtracking(string s, int start){
        if(start == s.length()) res.push_back(path);
        for(int i = start; i < s.length(); i++){
            if(isPalindrome(s, start, i)){
                path.push_back(s.substr(i, start-i+1));
                backtracking(s, i+1);
                path.pop_back();
            }
        }
    }
};

// 93复原IP地址
class Solution_93 {
private:
    vector<string> res;
public:
    void backtracking(string s, int startIndex, int tag, string path){
        if(startIndex >= s.length()){
            if(tag == 0){
                path.resize(path.length()-1);
                res.push_back(path);
            }
            return;
        }
        if(s[startIndex] == '0'){
            backtracking(s, startIndex+1, tag-1, path+'0'+'.');
        }
        else{
            for(int i = 1; i <= 3 && startIndex+i <= s.length(); i++){
                int singleIP = stoi(s.substr(startIndex, i));
                if(singleIP <= 255){
                    backtracking(s, startIndex+i, tag-1, path+s.substr(startIndex, i)+'.');
                }
            }
        }
    }
    vector<string> restoreIpAddresses(string s) {
        backtracking(s, 0, 4, "");
        return res;
    }
};

// 参考 想想为什么参考答案的效率这么高
class Solution_39_ref{
private:
    vector<string> res;
    bool isVaild(const string& s, int start, int end){
        if(start > end) return false;
        if(s[start] == '0' && start != end) return false;
        int num = 0;
        for(int i = start; i <= end; i++){
            if(s[i] > '9' || s[i] < '0') return false;
            num = num * 10 + (s[i] - '0');
            if(num > 255) return false;
        }
        return true;
    }
    void backtracking(string& s, int startIndex, int pointNum){
        if(pointNum == 3){
            if(isVaild(s, startIndex, s.size()-1)){
                res.push_back(s);
            }
            return;
        }
        for(int i = startIndex; i < s.size(); i++){
            if(isVaild(s, startIndex, i)){
                s.insert(s.begin()+i+1, '.');
                backtracking(s, i+2, pointNum+1);
                s.erase(s.begin()+i+1);
            }
            else break;
        }
    }
public:
    vector<string> restoreIpAddresses(string s) {
        res.clear();
        if (s.size() < 4 || s.size() > 12) return res; // 算是剪枝了
        backtracking(s, 0, 0);
        return res;
    }

};


class Solution_78 {
private:
    vector<vector<int>> res;
    vector<int> path;
public:
    void backtracking(int startIndex, vector<int>& nums){
        res.push_back(path);
        if(startIndex == nums.size()){
            return;
        }
        for(int i = startIndex; i < nums.size(); i++){
            path.push_back(nums[i]);
            backtracking(i+1, nums);
            path.pop_back();
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        backtracking(0, nums);
        return res;
    }
};


// 90 子集Ⅱ
class Solution_90 {
private:
    vector<vector<int>> res;
    vector<int> path;
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        backtracking(nums, 0);
        return res;
    }
    void backtracking(vector<int>& nums, int startIndex){
        res.push_back(path);
        if(startIndex >= nums.size()) return;
        for(int i = startIndex; i < nums.size(); i++){
            if(i > startIndex && nums[i] == nums[i-1]){
                continue;
            }
            path.push_back(nums[i]);
            backtracking(nums, i+1);
            path.pop_back();
        }
    }
};


// 非递减子序列
class Solution_491 {
private:
    vector<vector<int>> res;
    vector<int> path;
public:
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        backtracking(nums, 0);
        return res;
    }
    void backtracking(vector<int>& nums, int startIndex){
        if(path.size() >= 2) res.push_back(path);
        if(startIndex >= nums.size()) return;
        for(int i = startIndex; i < nums.size(); i++){
            if(i > startIndex && find(nums.begin()+startIndex, nums.begin()+i, nums[i]) != nums.begin()+i) 
                continue;
            if(path.empty() || nums[i] >= path[path.size()-1]){
                path.push_back(nums[i]);
                backtracking(nums, i+1);
                path.pop_back();
            }
        }
    }
};

class Solution_46 {
private:
    vector<vector<int>> res;
    vector<int> path;
public:
    void backtracking(vector<bool>& used, vector<int>& nums){
        if(path.size() == nums.size()){
            res.push_back(path);
            return;
        }
        for(int i = 0; i < nums.size(); i++){
            if(!used[i]){
                path.push_back(nums[i]);
                used[i] = true;
                backtracking(used, nums);
                used[i] = false;
                path.pop_back();
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<bool> used(nums.size(), false);
        backtracking(used, nums);
        return res;
    }
};

class Solution_47{
private:
    vector<vector<int>> res;
    vector<int> path;
public:
    void backtracking(vector<int>& nums, vector<bool>& used){
        if(path.size() == nums.size()){
            res.push_back(path);
            return;
        }
        for(int i = 0; i < nums.size(); i++){
            if(i != 0 && nums[i] == nums[i-1] && !used[i-1]) continue;
            if(used[i]) continue;
            path.push_back(nums[i]);
            used[i] = true;
            backtracking(nums, used);
            path.pop_back();
            used[i] = false;
        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<bool> used(nums.size(), false);
        backtracking(nums, used);
        return res;
    }

};

class Solution_332{
private:
    vector<string> res = {"JFK"};
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        vector<bool> used(tickets.size(), false);
        quicksort(tickets, 0, tickets.size()-1);
        backtracking(tickets, used, "JFK");
        return res;
    }

    void quicksort(vector<vector<string>>& tickets, int beginIndex, int endIndex){
        int left = beginIndex, right = endIndex;
        if(left >= right) return;
        int midTag = beginIndex;
        while(left < right){
            while(left < right && compare(tickets[midTag], tickets[right])) right--;
            if(left < right){
                swap(tickets[midTag], tickets[right]);
                midTag = right;
            }
            while(left < right && compare(tickets[left], tickets[midTag])) left++;
            if(left < right){
                swap(tickets[left], tickets[midTag]);
                midTag = left;
            }
        }
        quicksort(tickets, beginIndex, midTag-1);
        quicksort(tickets, midTag+1, endIndex);
    }

    bool compare(vector<string> a, vector<string> b){
        for(int i = 0; i < a[1].size(); i++){
            if(a[1].at(i) > b[1].at(i)) return false;
            else if(a[1].at(i) < b[1].at(i)) return true;
        }
        return true;
    }

    bool backtracking(vector<vector<string>>& tickets, vector<bool>& used, string place){
        if(res.size() == tickets.size()+1) return true;
        for(int i = 0; i < tickets.size(); i++){
            if(!used[i] && tickets[i][0] == place){
                used[i] = true;
                res.push_back(tickets[i][1]);
                if(backtracking(tickets, used, tickets[i][1])){
                    return true;
                }
                used[i] = false;
                res.pop_back();
            }
        }
        return false;
    }
};

class Solution_332_ref{
private:
    unordered_map<string, map<string, int>> targets;
    
    bool backtracking(int ticketNum, vector<string>& result){
        if(result.size() == ticketNum + 1){
            return true;
        }
        for(pair<const string ,int>& target: targets[result[result.size() - 1]]){
            if(target.second > 0){
                result.push_back(target.first);
                target.second--;
                if(backtracking(ticketNum, result)) return true;
                result.pop_back();
                target.second++;
            }
        }
        return false;
    }
public:
    vector<string> findItinerary(vector<vector<string>>& tickets){
        targets.clear();
        vector<string> result;
        for(const vector<string>& vec: tickets){
            targets[vec[0]][vec[1]]++;
        }
        result.push_back("JKF");
        backtracking(tickets.size(), result);
        return result;
    }
};

// 51 N皇后
class Solution_51 {
private:
    vector<vector<string>> res;
    vector<string> path;
public:
    vector<vector<string>> solveNQueens(int n) {
        path = vector<string> (n, string(n,'.'));
        backtracking(0, n);
        return res;
    }
    bool check(int row, int col, int n){
        // check line and rou
        for(int i = 0; i < n; i++){
            if(path[row][i] != '.' || path[i][col] != '.') return false;
        }
        // 对角线
        int diagonalRow = row - min(row, col);
        int diagonalCol = col - min(row, col);
        while(diagonalRow < n && diagonalCol < n){
            if(path[diagonalRow++][diagonalCol++] != '.') return false;
        }
        // 反对角线
        int antiDiagonalRow, antiDiagonalCol;
        if(row + col < n - 1){
            antiDiagonalRow = row + col;
            antiDiagonalCol = 0;
        }
        else{
            antiDiagonalRow = n - 1;
            antiDiagonalCol = col - (n - 1 - row);
        }
        while(antiDiagonalRow >= 0 && antiDiagonalCol < n){
            if(path[antiDiagonalRow--][antiDiagonalCol++] != '.') return false;
        }
        return true;
    }

    void backtracking(int row, int n){
        if(row == n){
            res.push_back(path);
            return;
        }
        for(int i = 0; i < n; i++){
            if(check(row, i, n)){
                path[row][i] = 'Q';
                backtracking(row+1, n);
                path[row][i] = '.';
            }
        }
    }

};

// 37 解数独
class Solution_37 {
public:
    void solveSudoku(vector<vector<char>>& board) {
        backtracking(board, 0, 0);
    }

    bool check(vector<vector<char>>& board, int row, int col, char put){
        // check row and col
        for(int i = 0; i < 9; i++){
            if(board[row][i] == put || board[i][col] == put) return false;
        }
        // check 3x3
        row -= row%3;
        col -= col%3;
        for(int i = 0; i < 3; i++){
            for(int j = 0; j < 3; j++){
                if(board[row+i][col+j] == put) return false;
            }
        }
        return true;
    }

    bool backtracking(vector<vector<char>>& board, int row, int col){
        if(col == 9){
            col = 0;
            row++;
        }
        if(row == 9){
            return true;
        }
        if(board[row][col]!='.'){
            return backtracking(board, row, col+1);
        }
        for(char ch = '1'; ch <= '9'; ch++){
            if(check(board, row, col, ch)){
                board[row][col] = ch;
                if(backtracking(board, row, col+1)) return true;
                board[row][col] = '.';
            }
        }
        return false;
    }
};

int main(){
    char a = '1';
    a++;
    cout<<a;
    Solution_51 s;
}